[Blogging Prompts Enhancements] Setting to enable/disable prompts

thomashorta opened the pull request:
Part of #17803
Depends on https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2645

Add a **Site Setting** to enable / disable the Blogging Prompts feature, which controls the visibility of the dashboard card and action in the "Create Post" Floating Action Button. Currently, the setting only shows up for sites that are considered potential blogs (this might be changed in future PRs), with the Blogging Prompt feature flag ON, AND with the Blogging Prompt Enhancements FF ON as well.

I the BP Enhancements FF is off, then we consider only the original conditions for the feature to show up (Blogging Prompt FF ON and the site is a potential blog). The only change here (even without the FF) is that we no longer check for the prompt being included in the Blogging Reminders, as Prompts and Reminders are separate features and this Reminder-specific setting should not interfere with the Prompts feature visibility. (internal ref: p1675113452461219-slack-C01CW1VMLAF)

Demo:

https://user-images.githubusercontent.com/5091503/215863093-8bbd2929-7d33-48bc-9cf9-eeb84101b600.mp4


## To test:

### Setting only shows up with Enhancements FF on
1. Install Jetpack and log in
2. Open "Debug settings" and make sure `BloggingPromptsEnhancementsFeatureConfig` is disabled
3. Open "My Site" -> "MENU" -> "Site Settings" with a blog selected
7. **Verify** the `Blogging Prompts` setting is NOT available in the `General` section
8. Go back
9. Open "Debug settings" (avatar on top right -> App Settings -> Debug Settings) and enable `BloggingPromptsEnhancementsFeatureConfig` in "Features in Development" section
10. Tap the "Restart the app" button
3. Open "My Site" -> "MENU" -> "Site Settings" with a blog selected
7. **Verify** the `Blogging Prompts` setting is available in the `General` section

### Dashboard Card and FAB action shows up with Enhancements FF off
1. Install Jetpack and log in
2. Open "Debug settings" and make sure `BloggingPromptsEnhancementsFeatureConfig` is disabled
3. Choose a site that is a blog
7. **Verify** the `Blogging Prompts` card is available in the dashboard
8. Tap the FAB to create a post
9. **Verify** the `Blogging Prompts` action is available in the bottom sheet

### Dashboard Card and FAB action only shows up when BP setting is ON if Enhancements FF is ON
1. Install Jetpack and log in
3. Open "Debug settings" and enable `BloggingPromptsEnhancementsFeatureConfig` in "Features in Development" section
4. Tap the "Restart the app" button
5. Open "My Site" -> "MENU" -> "Site Settings" with a blog selected
6. Make sure the `Blogging Prompts` setting is ON in the `General` section
7. Go back to "My Site" -> "HOME"
8. **Verify** the `Blogging Prompts` card is available in the dashboard
9. Tap the FAB to create a post
10. **Verify** the `Blogging Prompts` action is available in the bottom sheet
11. Open "My Site" -> "MENU" -> "Site Settings" with a blog selected
12. Turn off the `Blogging Prompts` setting in the `General` section
7. Go back to "My Site" -> "HOME"
8. **Verify** the `Blogging Prompts` card is NOT available in the dashboard
9. Tap the FAB to create a post
10. **Verify** the `Blogging Prompts` action is NOT available in the bottom sheet

If possible, try going through the tests above selecting a site that is NOT a blog. The Prompts feature and setting should NOT be available at all on that type of site.

## Regression Notes
1. Potential unintended areas of impact
Unexpected behavior changes with the Enhancements Flag off, tracking for the Prompts card viewed being triggered even when the user disables it.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

7. What automated tests I added (or what prevented me from doing so)
Required unit tests were updated and new tests were added, especially for the new `BloggingPromptsSettingsHelper`.

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
add Prompts enabled setting in Site Settings

Commit:
point to FluxC PR version

Commit:
move prompts active setting to helper and use it in dashboard

Commit:
only show prompts setting if prompts availability conditions are met

The Blogging Prompts setting should only be shown in the Site Settings if the
feature is available for the current site. Which means the site and user has to
meet some other conditions, and not only the Prompts feature flag being on.

Commit:
also check for prompts being active in "create site" bottom sheet

We should not display the Blogging Prompts action in the "Create Site" bottom
sheet if the user has disabled the Prompts in the "Site Settings" as well as if
the card would not be visible because of other conditions (ex: was skipped for
today). To do so, we should use the `isPromptsFeatureActive` function to make
sure all conditions are checked.

Commit:
update existing unit tests

Commit:
don't check prompts reminder opt-int for prompts availability

We discussed internally and now that we have a excluive setting for the Blogging
Prompts, we should not be looking at the Prompts being enabled for Reminders
to decide if the dashboard / FAB action should be shown.

Commit:
add unit tests for BP settings helper

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsSettingsHelper.kt:
```diff
@@ -0,0 +1,68 @@
+        updatePromptsCardEnabled(siteId, isEnabled)
+    }
+
+    @Suppress("MemberVisibilityCanBePrivate")
```
I was wondering if we really need this suppress annotation.

thomashorta replied:
Nope, I forgot to remove it. Thanks for catching that!

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsSettingsHelper.kt:
```diff
@@ -0,0 +1,68 @@
+        return bloggingPromptsFeatureConfig.isEnabled() && selectedSite.isPotentialBloggingSite
+    }
+
+    suspend fun isPromptsFeatureActive(): Boolean {
```
To me `Available` (`isPromptsFeatureAvailable`) and `Active` (`isPromptsFeatureActive`) sound the same in this context. I was trying to think of a name suggestion, but couldn't think of one 🤔

thomashorta replied:
Yeah, I got those confused more than once while developing. I could only think of the following:
- `isPromptsFeatureAvailable` remains the same
- `isPromptsFeatureActive` -> `shouldShowPromptsFeature`

Wdyt?

RenanLukas replied:
Makes sense to me 👍

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java:
```diff
@@ -1239,12 +1248,14 @@ private void initBloggingReminders() {
+                    .getBlogSettingsUiState(mSite.getId())
+                    .observe(getAppCompatActivity(), s -> {
+                        if (mBloggingRemindersPref != null) {
+                            CharSequence summary = mUiHelpers.getTextOfUiString(getActivity(), s);
```
Nitpick: would it make sense to make it `final`?

RenanLukas commented on WordPress/src/test/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsSettingsHelperTest.kt:
```diff
@@ -0,0 +1,294 @@
+    fun `given prompts FF is off and site is potential blog, when isPromptsFeatureAvailable, then returns false`() {
+        whenever(bloggingPromptsFeatureConfig.isEnabled()).thenReturn(false)
+        whenever(selectedSiteRepository.getSelectedSite()).thenReturn(
+            SiteModel().apply { setIsPotentialBloggingSite(true) }
```
I was wondering if it makes sense to extract the code creating a `SiteModel` instance that repeats itself to a private method in this class 🤔

RenanLukas reviewed (commented):
Thanks for the PR, @thomashorta ! It works as expected 🎉 

I've left a few minor questions/comments. Please let me know if the make sense.

One minor thing I've noticed while testing was that the FAB prompts UI was shown before it was quickly hidden after disabling the prompts in site settings. Steps to reproduce:

1 - Enable the blogging prompts in `Site settings`;
2 - Tap the FAB in home: prompts UI is shown, as expected;
3 - Disable the blogging prompts in `Site settings`;
4 - Tap the FAB in home: prompts UI is quickly shown before it is hidden.

I believe that happens because we keep the same bottom sheet UI instance.

Commit:
use postValue instead of postState in CardSource

I noticed that sometimes the progress spinner from the home cards list was
not hiding even though the cards were fully loaded. I checked that this happened
when I started using postState in the BloggingPromptCardSource#getData, so I am
"reverting" that change to use postValue as it was before.

Commit:
remove unneeded annotation

Commit:
fix bottom sheet content shifting around when displayed

Commit:
rename isPropmtsFeatureActive to shouldShowPromptsFeature

This was done for better clarity since "Active" and "Available" are very similar

Commit:
make java variable final

Commit:
extract model creation logic in helper tests

thomashorta commented:
> One minor thing I've noticed while testing was that the FAB prompts UI was shown before it was quickly hidden after disabling the prompts in site settings.

Nice! I saw that issue as well and forgot to look into it 😓 

I just fixed it by preventing the bottom sheet to show up until its content is loaded, we were loading the content AND showing the bottom sheet at the same time, so it was being shown with stale content for a brief moment, as you observed. Thanks for catching that!

---

I also fixed (hopefully) another issue I was seeing after this change, which was the loading indicator never hiding in My Site Home even after all cards finished loading. The fix was revert to using `postValue` instead of `postState` in `BloggingPromptCardSource#getData`

RenanLukas reviewed (approved):
Thanks for the changes, @thomashorta . It's working as expected :shipit:

Commit:
update fluxC dependency to point to a trunk commit
